Sources/ClientRuntime/Networking/Http/URI.swift

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+import Foundation
+
+public struct URI: Hashable {
+    public let scheme: Scheme
+    public let path: String
+    public let host: String
+    public let port: Int16
+    public let query: [SDKURLQueryItem]
+    public let username: String?
+    public let password: String?
+    public var url: URL? {
+        self.toBuilder().getUrl()
+    }
+    public var queryString: String? {
+        if self.query.isEmpty {
+            return nil
+        }
+
+        return self.query.map { queryItem in
+            return [queryItem.name, queryItem.value].compactMap { $0 }.joined(separator: "=")
+        }.joined(separator: "&")
+    }
+
+    fileprivate init(scheme: Scheme,
+                     path: String,
+                     host: String,
+                     port: Int16,
+                     query: [SDKURLQueryItem],
+                     username: String? = nil,
+                     password: String? = nil) {
+        self.scheme = scheme
+        self.path = path
+        self.host = host
+        self.port = port
+        self.query = query
+        self.username = username
+        self.password = password
+    }
+
+    public func toBuilder() -> URIBuilder {
+        return URIBuilder()
+           .withScheme(self.scheme)
+           .withPath(self.path)
+           .withHost(self.host)
+           .withPort(self.port)
+           .withQueryItems(self.query)
+    }
+}
+
+public class URIBuilder {
+    var urlComponents: URLComponents
+    var scheme: Scheme = Scheme.https
+    var path: String = "/"
+    var host: String = ""
+    var port: Int16 = Int16(Scheme.https.port)
+    var query: [SDKURLQueryItem] = []
+    var username: String?
+    var password: String?
+
+    required public init() {
+        self.urlComponents = URLComponents()
+        self.urlComponents.scheme = self.scheme.rawValue
+        self.urlComponents.path = self.path
+        self.urlComponents.host = self.host
+    }
+
+    @discardableResult
+    public func withScheme(_ value: Scheme?) -> URIBuilder {
+        self.scheme = value ?? Scheme.https
+        self.urlComponents.scheme = self.scheme.rawValue
+        return self
+    }
+
+    @discardableResult
+    public func withPath(_ value: String) -> URIBuilder {
+        self.path = value
+        if self.path.contains("%") {
+            self.urlComponents.percentEncodedPath = self.path
+        } else {
+            self.urlComponents.path = self.path
+        }
```

Comment:
Instead of this check, you could just allow setting percent-encoded path and normal path, like URLComponents does

Reply:
A consequence of this though is that callers of `URI.path` now can't know whether the `URI` is percent encoded or not. I guess they use the String extension you added though. But if they wanted it to be encoded and it's not, they have to do it themselves?